File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/root/TransactionCategoriesRootMVP.java
package com.androidcollider.easyfin.transaction_categories.root;

import androidx.annotation.Nullable;
import androidx.core.util.Pair;

import com.androidcollider.easyfin.common.models.TransactionCategory;

import java.util.List;

import io.reactivex.Flowable;

/**
 * @author Ihor Bilous
 */

public interface TransactionCategoriesRootMVP {

    interface Model {

        Flowable<Pair<List<TransactionCategory>, List<TransactionCategory>>> getAllTransactionCategories();

        Flowable<TransactionCategory> addNewTransactionCategory(TransactionCategory transactionCategory, boolean isExpense);
    }

    interface View {

        void showMessage(String message);

        void shakeDialogNewTransactionCategoryField();

        void dismissDialogNewTransactionCategory();

        void handleNewTransactionCategoryAdded();
    }

    interface Presenter {

        void setView(@Nullable TransactionCategoriesRootMVP.View view);

        void loadData();

        void addNewCategory(String name, boolean isExpense);
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/root/TransactionCategoriesRootMVP.kt
package com.androidcollider.easyfin.transaction_categories.root

import com.androidcollider.easyfin.common.models.TransactionCategory
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
interface TransactionCategoriesRootMVP {
    interface Model {
        val allTransactionCategories:
                Single<Pair<List<TransactionCategory>, List<TransactionCategory>>>

        fun addNewTransactionCategory(
            transactionCategory: TransactionCategory,
            isExpense: Boolean
        ): Single<TransactionCategory>
    }

    interface View {
        fun showMessage(message: String)
        fun shakeDialogNewTransactionCategoryField()
        fun dismissDialogNewTransactionCategory()
        fun handleNewTransactionCategoryAdded()
    }

    interface Presenter {
        fun setView(view: View?)
        fun loadData()
        fun addNewCategory(name: String, isExpense: Boolean)
    }
}